src/node.cc

```diff
@@ -1024,12 +1024,39 @@ InitializationResult InitializeOncePerProcess(int argc, char** argv) {
   // In the case of FIPS builds we should make sure
   // the random source is properly initialized first.
 #if OPENSSL_VERSION_MAJOR >= 3
-  if (EVP_default_properties_is_fips_enabled(nullptr)) {
+  // Call OPENSSL_init_crypto to initialize OPENSSL_INIT_LOAD_CONFIG to
+  // avoid the default behavior where errors raised during the parsing of the
+  // OpenSSL configuration file are not propagated and cannot be detected.
+  //
+  // If FIPS is configured the OpenSSL configuration file will have an .include
+  // pointing to the fipsmodule.cnf file generated by the openssl fipsinstall
+  // command. If the path to this file is incorrect no error will be reported.
+  // For Node.js this will mean that EntropySource will be called by V8 as part
+  // of its initalization process, and EntropySource will in turn call
+  // CheckEntropy. CheckEntropy will call RAND_status which will now always
+  // return 0 leading to an endless loop and the node process will appear to
+  // hang/freeze.
+  std::string openssl_conf;
+  credentials::SafeGetenv("OPENSSL_CONF", &openssl_conf);
+  if (!per_process::cli_options->openssl_config.empty() ||
+      !openssl_conf.empty()) {
+    OPENSSL_INIT_SETTINGS* settings = OPENSSL_INIT_new();
+    OPENSSL_INIT_set_config_file_flags(settings, CONF_MFLAGS_DEFAULT_SECTION);
+    OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, settings);
```

Comment:
Do we need to call `OPENSSL_INIT_set_config_filename()` with the name of the config file if set via the env var or cli flag? Similar to how we do in
https://github.com/nodejs/node/blob/0d7644fddaba7979b08805c6c8fff4ab0677a72e/src/crypto/crypto_util.cc#L108-L121
?

And come to think of it, is the code being modified here called before InitCryptoOnce?

Reply:
Ah I think we don't need to call `OPENSSL_INIT_set_config_filename` for the environment variable case, as OpenSSL already knows about `OPENSSL_CONFIG` and handles that case, but we do appear to need to if we set the cli flag, which is our own and OpenSSL has no in-built knowledge about.

Reply:
Good point about `OPENSSL_CONF`, I did not think of that and have set it in the latest commit. I'll remove this though. Thanks again.